vnext/Shared/Networking/OriginPolicyHttpFilter.cpp

```diff
@@ -0,0 +1,638 @@
+#include <RuntimeOptions.h>
+#include "WinRTTypes.h"
+
+// Boost Library
+#include <boost/algorithm/string.hpp>
+#include <boost/lexical_cast/try_lexical_convert.hpp>
+
+// Windows API
+#include <winrt/Windows.Web.Http.Headers.h>
+
+// Standard Library
+#include <regex>
+
+using std::set;
+using std::wstring;
+
+using winrt::hresult_error;
+using winrt::hstring;
+using winrt::to_hstring;
+using winrt::Windows::Foundation::IInspectable;
+using winrt::Windows::Foundation::Uri;
+using winrt::Windows::Web::Http::HttpMethod;
+using winrt::Windows::Web::Http::HttpRequestMessage;
+using winrt::Windows::Web::Http::HttpResponseMessage;
+using winrt::Windows::Web::Http::Filters::IHttpFilter;
+using winrt::Windows::Web::Http::Headers::HttpMediaTypeHeaderValue;
+using winrt::Windows::Web::Http::Headers::HttpRequestHeaderCollection;
+
+namespace Microsoft::React::Networking {
+
+#pragma region OriginPolicyHttpFilter
+
+// https://fetch.spec.whatwg.org/#forbidden-method
+/*static*/ set<const wchar_t *> OriginPolicyHttpFilter::s_forbiddenMethods = {L"CONNECT", L"TRACE", L"TRACK"};
+
+/*static*/ set<const wchar_t *> OriginPolicyHttpFilter::s_simpleCorsMethods = {L"GET", L"HEAD", L"POST"};
+
+/*static*/ set<const wchar_t *> OriginPolicyHttpFilter::s_simpleCorsRequestHeaderNames = {
+    L"Accept",
+    L"Accept-Language",
+    L"Content-Language",
+    L"Content-Type",
+    L"DPR",
+    L"Downlink",
+    L"Save-Data",
+    L"Viewport-Width",
+    L"Width"};
+
+/*static*/ set<const wchar_t *> OriginPolicyHttpFilter::s_simpleCorsResponseHeaderNames =
+    {L"Cache-Control", L"Content-Language", L"Content-Type", L"Expires", L"Last-Modified", L"Pragma"};
+
+/*static*/ set<const wchar_t *> OriginPolicyHttpFilter::s_simpleCorsContentTypeValues = {
+    L"application/x-www-form-urlencoded",
+    L"multipart/form-data",
+    L"text/plain"};
+
+// https://fetch.spec.whatwg.org/#forbidden-header-name
+// Chromium still bans "User-Agent" due to https://crbug.com/571722 //TODO: Remove?
+/*static*/ set<const wchar_t *> OriginPolicyHttpFilter::s_corsForbiddenRequestHeaderNames = {
+    L"Accept-Charset",
+    L"Accept-Encoding",
+    L"Access-Control-Request-Headers",
+    L"Access-Control-Request-Method",
+    L"Connection",
+    L"Content-Length",
+    L"Cookie",
+    L"Cookie2",
+    L"Date",
+    L"DNT",
+    L"Expect",
+    L"Host",
+    L"Keep-Alive",
+    L"Origin",
+    L"Referer",
+    L"TE",
+    L"Trailer",
+    L"Transfer-Encoding",
+    L"Upgrade",
+    L"Via"};
+
+/*static*/ set<const wchar_t *> OriginPolicyHttpFilter::s_corsForbiddenRequestHeaderNamePrefixes = {L"Proxy-", L"Sec-"};
+
+/*static*/ Uri OriginPolicyHttpFilter::s_origin{nullptr};
+
+/*static*/ void OriginPolicyHttpFilter::SetStaticOrigin(const char *url) {
+  if (url)
+    s_origin = Uri{to_hstring(url)};
+}
+
+/*static*/ bool OriginPolicyHttpFilter::IsSameOrigin(Uri const &u1, Uri const &u2) noexcept {
+  return u1.SchemeName() == u2.SchemeName() && u1.Host() == u2.Host() && u1.Port() == u2.Port();
```

Comment:
>u1.Host() == u2.Host() [](http://example.com/codeflow?start=47&length=22)

The doc for Host states:

> If the domain name is subject to Punycode encoding as shown in AbsoluteCanonicalUri, the Host property shows the decoded value, not the Punycode encoded value.

Just wanted to doublecheck that this is okay here to compare the strings directly, since there might be multiple encoded host values mapping to the same decoded URL...